File: quicksort-13b.java

public class Quicksort {
    public static void main(String[] args) {
        int[] arr = {5, 1, 3, 2, 4};
        quicksort(arr, 0, 4);
        for (int i = 0; i < 5; i++) {
            System.out.println(arr[i]);
        }
    }

    public static void quicksort(int[] arr, int low, int high) {
        if (low < high) {
            int partition = partition(arr, low, high);
            quicksort(arr, low, partition - 1);
            quicksort(arr, partition + 1, high);
        }
    }

    public static int partition(int[] arr, int low, int high) {
        int pivot = arr[low];
        int i = low;
        for (int j = low + 1; j <= high; j++) {
            if (arr
[j] <= pivot) {
                i++;
                swap(arr, i, j);
            }
        }
        swap(arr, low, i);
        return i;
    }

    public static void swap(int[] arr, int i, int j) {
        int tmp = arr[i];
        arr[i] = arr[j];
        arr[j] = tmp;
    }
}

File: quicksort-13b.c
#include <stdio.h>
#include <stdlib.h>

void quicksort(int *arr, int left, int right)
{
    int i, j, tmp, pivot;
    if (left >= right) {
        return;
    }
    pivot = arr[left];
    i = left;
    j = right;
    while (i < j) {
        while (i < j && arr[j] >= pivot) {
            j--;
        }
        arr[i] = arr[j];
        while (i < j && arr[i] <= pivot) {
            i++;
        }
        arr[j] = arr[i];
    }
    arr[i] = pivot;
    quicksort(arr, left, i - 1);
    quicksort(arr, i + 1, right);
}

int main()
{
    int arr[] = {3, 2, 1, 5, 6, 4};
    int len = sizeof(arr) / sizeof(arr[0]);

    quicksort(arr, 0, len - 1);
    for (int i = 0; i < len; i++) {
        printf("%d ", arr[i]);
    }
    return 0;
}
